app/controllers/submissions_controller.rb

```diff
@@ -466,6 +471,40 @@ def download
     end
   end
 
+  def notebook_content
+    if params[:select_file_id]
+      file = SubmissionFile.find(params[:select_file_id])
+      file_contents = file.retrieve_file
+      grouping = file.submission.grouping
+      assignment = grouping.assignment
+      path = Pathname.new(file.path).relative_path_from(Pathname.new(assignment.repository_folder)).to_s
+      revision_identifier = file.submission.revision_identifier
+      filename = file.filename
+    else
+      assignment = Assignment.find(params[:assignment_id])
+      grouping = find_appropriate_grouping(assignment.id, params)
+      revision_identifier = params[:revision_identifier]
+
+      path = params[:path] || '/'
+      file_contents = grouping.access_repo do |repo|
+        if revision_identifier.nil?
+          revision = repo.get_latest_revision
+        else
+          revision = repo.get_revision(revision_identifier)
+        end
+        file = revision.files_at_path(File.join(assignment.repository_folder, path))[params[:file_name]]
+        repo.download_as_string(file)
+      end
+      filename = params[:file_name]
+    end
+
+    file_path = "#{assignment.repository_folder}/#{path}/#{filename}"
+    unique_path = "#{grouping.group.repo_name}/#{file_path}.#{revision_identifier}"
+    @notebook_type = File.extname(filename).casecmp('.ipynb')&.zero? ? 'ipynb' : 'rmarkdown'
```

Comment:
I think this would be a great place for `FileHelper.get_file_type`. If you want you could also add a check for file types that aren't `ipynb`/`rmarkdown`, and render different content in that case?